runtime-transaction/src/instruction_details.rs

```diff
@@ -0,0 +1,355 @@
+use {
+    solana_builtins_default_costs::MAYBE_BUILTIN_KEY,
+    solana_compute_budget::compute_budget_limits::*,
+    solana_sdk::{
+        instruction::{CompiledInstruction, InstructionError},
+        pubkey::Pubkey,
+        saturating_add_assign,
+        transaction::{Result, TransactionError},
+    },
+    std::num::NonZeroU32,
+};
+
+/// Information about instructions gathered after scan over transaction;
+/// These are "raw" information that suitable for cache and reuse.
+#[cfg_attr(test, derive(Eq, PartialEq))]
+#[derive(Default, Debug)]
+pub struct InstructionDetails {
+    // compute-budget instruction details:
+    // the first field in tuple is instruction index, second field is the unsanitized value set by user
+    pub requested_compute_unit_limit: Option<(u8, u32)>,
+    pub requested_compute_unit_price: Option<(u8, u64)>,
+    pub requested_heap_size: Option<(u8, u32)>,
+    pub requested_loaded_accounts_data_size_limit: Option<(u8, u32)>,
+    // builtin instruction details
+    pub sum_builtin_compute_units: u32,
+    pub count_builtin_instructions: u32,
+    pub count_non_builtin_instructions: u32,
+    pub count_compute_budget_instructions: u32,
+}
+
+impl InstructionDetails {
+    pub fn try_from<'a>(
+        instructions: impl Iterator<Item = (&'a Pubkey, &'a CompiledInstruction)>,
+    ) -> Result<Self> {
+        let mut instruction_details = InstructionDetails::default();
+
+        for (i, (program_id, instruction)) in instructions.enumerate() {
+            if MAYBE_BUILTIN_KEY[program_id.to_bytes()[0] as usize] {
```

Comment:
Use `as_ref` to avoid the copy
```suggestion
            if MAYBE_BUILTIN_KEY[program_id.as_ref()[0] as usize] {
```